File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Managers/NavigationManager.swift
import Combine
import SwiftUI

// Model for category navigation parameters
struct CategoryNavigationParams: Equatable {
    let categoryId: String?
    let categoryName: String?
    let subCategoryId: String?
    let filters: [String: String]
    let layoutType: String?

    init(
        categoryId: String? = nil, categoryName: String? = nil, subCategoryId: String? = nil,
        filters: [String: String] = [:], layoutType: String? = nil
    ) {
        self.categoryId = categoryId
        self.categoryName = categoryName
        self.subCategoryId = subCategoryId
        self.filters = filters
        self.layoutType = layoutType
    }
}

enum TabType: String, CaseIterable, Identifiable {
    case shopping = "Shopping"
    case services = "Services"
    case grocery = "Grocery"
    case influencers = "Influencers"

    var id: String { rawValue }

    var iconName: String {
        switch self {
        case .shopping: return "bag.fill"
        case .services: return "building.2.fill"
        case .grocery: return "basket.fill"
        case .influencers: return "person.3.fill"
        }
    }

    var color: Color {
        switch self {
        // Using System Colors for now to avoid init errors until Extension is confirmed
        case .shopping: return Color(hex: "#2563EB")
        case .services: return Color(hex: "#7C3AED")
        case .grocery: return Color(hex: "#10B981")
        case .influencers: return Color(hex: "#EC4899")
        }
    }
}

// Enum for overlay navigation — replaces 7 separate @Published booleans
enum OverlayDestination: String, Identifiable {
    case beauty
    case specialDeal
    case brandNewArrival
    case menFashion
    case grandMobiles
    case shoesSales
    case cyberSale
    case categoryPage

    var id: String { rawValue }
}

class NavigationManager: ObservableObject {
    @Published var selectedCategory: String = "For You"
    @Published var activeTab: TabType = .shopping

    // Single enum replaces 7+ booleans — one objectWillChange instead of many
    @Published var activeOverlay: OverlayDestination? = nil

    // Category navigation state
    @Published var categoryNavigation: CategoryNavigationParams?

    // Computed compatibility properties for existing code
    var showBeautyPage: Bool {
        get { activeOverlay == .beauty }
        set { activeOverlay = newValue ? .beauty : nil }
    }
    var showSpecialDealPage: Bool {
        get { activeOverlay == .specialDeal }
        set { activeOverlay = newValue ? .specialDeal : nil }
    }
    var showBrandNewArrivalPage: Bool {
        get { activeOverlay == .brandNewArrival }
        set { activeOverlay = newValue ? .brandNewArrival : nil }
    }
    var showMenFashionPage: Bool {
        get { activeOverlay == .menFashion }
        set { activeOverlay = newValue ? .menFashion : nil }
    }
    var showGrandMobilesPage: Bool {
        get { activeOverlay == .grandMobiles }
        set { activeOverlay = newValue ? .grandMobiles : nil }
    }
    var showShoesSalesPage: Bool {
        get { activeOverlay == .shoesSales }
        set { activeOverlay = newValue ? .shoesSales : nil }
    }
    var showCyberSalePage: Bool {
        get { activeOverlay == .cyberSale }
        set { activeOverlay = newValue ? .cyberSale : nil }
    }
    var showCategoryPage: Bool {
        get { activeOverlay == .categoryPage }
        set { activeOverlay = newValue ? .categoryPage : nil }
    }

    func navigate(to url: String) {
        // Handle grocery category deep links with multiple IDs
        if url.starts(with: "/grocery/category/") {
            let ids = url.replacingOccurrences(of: "/grocery/category/", with: "")
            self.categoryNavigation = CategoryNavigationParams(
                categoryId: nil,  // Category ID might not be needed if we have sub-category IDs
                categoryName: "Groceries",
                subCategoryId: ids,
                layoutType: "grocery"
            )
            self.showCategoryPage = true
            return
        }

        // Handle category:// URLs
        if url.starts(with: "category://") {
            parseCategoryURL(url)
            return
        }

        // Handle slug-based navigation
        if url == "shopping" {
            if self.activeTab == .shopping {
                self.selectedCategory = "For You"
            }
            self.activeTab = .shopping
        } else if url == "services" {
            self.activeTab = .services
        } else if url == "grocery" {
            self.activeTab = .grocery
        } else if url == "influencers" {
            self.activeTab = .influencers
        } else if url == "beauty-product" {
            self.showBeautyPage = true
        } else if url == "special-deal-new-style" {
            self.showSpecialDealPage = true
        } else if url == "brand-new-arrival" {
            self.showBrandNewArrivalPage = true
        } else if url == "men-fashion" {
            self.showMenFashionPage = true
        } else if url == "grand-mobiles-sale" {
            self.showGrandMobilesPage = true
        } else if url == "footwear-collection" || url == "footwear-sale-collection" {
            self.showShoesSalesPage = true
        } else if url == "cyber-sale" {
            self.showCyberSalePage = true
        } else if url.starts(with: "/collection/") {
            AppLogger.debug("Navigate to collection: \(url)")
        } else {
            AppLogger.debug("Unhandled navigation: \(url)")
        }
    }

    // Parse category:// URLs
    // Format: category://categoryName?categoryId=xxx&subCategoryId=yyy&filters=key:value
    private func parseCategoryURL(_ url: String) {
        // Remove the scheme
        let withoutScheme = url.replacingOccurrences(of: "category://", with: "")

        // Split path and query
        let components = withoutScheme.components(separatedBy: "?")
        let categoryName = components.first?.removingPercentEncoding ?? "Category"

        var categoryId: String?
        var subCategoryId: String?
        var layoutType: String?
        var filters: [String: String] = [:]

        // Parse query params
        if components.count > 1 {
            let queryString = components[1]
            let params = queryString.components(separatedBy: "&")

            for param in params {
                let keyValue = param.components(separatedBy: "=")
                guard keyValue.count == 2 else { continue }

                let key = keyValue[0]
                let value = keyValue[1].removingPercentEncoding ?? keyValue[1]

                switch key {
                case "categoryId":
                    categoryId = value
                case "subCategoryId":
                    subCategoryId = value
                case "layout":
                    layoutType = value
                case "filters":
                    // Parse filters format: key1:value1,key2:value2
                    let filterPairs = value.components(separatedBy: ",")
                    for pair in filterPairs {
                        let kv = pair.components(separatedBy: ":")
                        if kv.count == 2 {
                            filters[kv[0]] = kv[1]
                        }
                    }
                default:
                    break
                }
            }
        }

        // Set navigation state
        self.categoryNavigation = CategoryNavigationParams(
            categoryId: categoryId,
            categoryName: categoryName,
            subCategoryId: subCategoryId,
            filters: filters,
            layoutType: layoutType
        )

        self.showCategoryPage = true
    }

    func dismissCategoryPage() {
        self.showCategoryPage = false
        self.categoryNavigation = nil
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/data/manager/NavigationManager.kt
package com.ecommerceearn.app.data.manager

import androidx.compose.ui.graphics.Color
import com.ecommerceearn.app.utils.AppLogger
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

data class CategoryNavigationParams(
    val categoryId: String? = null,
    val categoryName: String? = null,
    val subCategoryId: String? = null,
    val filters: Map<String, String> = emptyMap(),
    val layoutType: String? = null
)

enum class TabType(val id: String, val iconName: String, val colorHex: String) {
    SHOPPING("Shopping", "bag.fill", "#2563EB"),
    SERVICES("Services", "building.2.fill", "#7C3AED"),
    GROCERY("Grocery", "basket.fill", "#10B981"),
    INFLUENCERS("Influencers", "person.3.fill", "#EC4899");
}

enum class OverlayDestination(val id: String) {
    BEAUTY("beauty"),
    SPECIAL_DEAL("specialDeal"),
    BRAND_NEW_ARRIVAL("brandNewArrival"),
    MEN_FASHION("menFashion"),
    GRAND_MOBILES("grandMobiles"),
    SHOES_SALES("shoesSales"),
    CYBER_SALE("cyberSale"),
    CATEGORY_PAGE("categoryPage")
}

enum class MainTab {
    HOME, CATEGORIES, CART, ACCOUNT
}

object NavigationManager {
    private val _selectedCategory = MutableStateFlow("For You")
    val selectedCategory: StateFlow<String> = _selectedCategory.asStateFlow()

    private val _activeTab = MutableStateFlow(MainTab.HOME)
    val activeTab: StateFlow<MainTab> = _activeTab.asStateFlow()

    private val _activeOverlay = MutableStateFlow<OverlayDestination?>(null)
    val activeOverlay: StateFlow<OverlayDestination?> = _activeOverlay.asStateFlow()

    private val _categoryNavigation = MutableStateFlow<CategoryNavigationParams?>(null)
    val categoryNavigation: StateFlow<CategoryNavigationParams?> = _categoryNavigation.asStateFlow()

    private val _groceryProductId = MutableStateFlow<String?>(null)
    val groceryProductId: StateFlow<String?> = _groceryProductId.asStateFlow()

    private val _isGroceryTabActive = MutableStateFlow(false)
    val isGroceryTabActive: StateFlow<Boolean> = _isGroceryTabActive.asStateFlow()

    private val _isServicesTabActive = MutableStateFlow(false)
    val isServicesTabActive: StateFlow<Boolean> = _isServicesTabActive.asStateFlow()

    private val _isInfluencersTabActive = MutableStateFlow(false)
    val isInfluencersTabActive: StateFlow<Boolean> = _isInfluencersTabActive.asStateFlow()

    fun setGroceryTabActive(active: Boolean) {
        _isGroceryTabActive.value = active
    }

    fun setServicesTabActive(active: Boolean) {
        _isServicesTabActive.value = active
    }

    fun setInfluencersTabActive(active: Boolean) {
        _isInfluencersTabActive.value = active
    }

    fun openGroceryProduct(id: String) {
        _groceryProductId.value = id
    }

    fun dismissGroceryProduct() {
        _groceryProductId.value = null
    }

    // Helper visibility states based on overlay
    val showBeautyPage: Boolean get() = _activeOverlay.value == OverlayDestination.BEAUTY
    val showSpecialDealPage: Boolean get() = _activeOverlay.value == OverlayDestination.SPECIAL_DEAL
    val showBrandNewArrivalPage: Boolean get() = _activeOverlay.value == OverlayDestination.BRAND_NEW_ARRIVAL
    val showMenFashionPage: Boolean get() = _activeOverlay.value == OverlayDestination.MEN_FASHION
    val showGrandMobilesPage: Boolean get() = _activeOverlay.value == OverlayDestination.GRAND_MOBILES
    val showShoesSalesPage: Boolean get() = _activeOverlay.value == OverlayDestination.SHOES_SALES
    val showCyberSalePage: Boolean get() = _activeOverlay.value == OverlayDestination.CYBER_SALE
    val showCategoryPage: Boolean get() = _activeOverlay.value == OverlayDestination.CATEGORY_PAGE

    fun setOverlay(destination: OverlayDestination?) {
        _activeOverlay.value = destination
    }

    fun navigate(url: String) {
        if (url.startsWith("/grocery/category/")) {
            val ids = url.replace("/grocery/category/", "")
            _categoryNavigation.value = CategoryNavigationParams(
                categoryName = "Groceries",
                subCategoryId = ids,
                layoutType = "grocery"
            )
            setOverlay(OverlayDestination.CATEGORY_PAGE)
            return
        }

        if (url.startsWith("category://")) {
            parseCategoryURL(url)
            return
        }

        when (url) {
            "shopping", "home" -> {
                _activeTab.value = MainTab.HOME
            }
            "categories" -> _activeTab.value = MainTab.CATEGORIES
            "cart" -> _activeTab.value = MainTab.CART
            "account" -> _activeTab.value = MainTab.ACCOUNT
            
            // Legacy internal routings
            "services" -> AppLogger.debug("Services requested from within Home Tab")
            "grocery" -> AppLogger.debug("Grocery requested from within Home Tab")
            "influencers" -> AppLogger.debug("Influencers requested from within Home Tab")
            
            "beauty-product" -> setOverlay(OverlayDestination.BEAUTY)
            "special-deal-new-style" -> setOverlay(OverlayDestination.SPECIAL_DEAL)
            "brand-new-arrival" -> setOverlay(OverlayDestination.BRAND_NEW_ARRIVAL)
            "men-fashion" -> setOverlay(OverlayDestination.MEN_FASHION)
            "grand-mobiles-sale" -> setOverlay(OverlayDestination.GRAND_MOBILES)
            "footwear-collection", "footwear-sale-collection" -> setOverlay(OverlayDestination.SHOES_SALES)
            "cyber-sale" -> setOverlay(OverlayDestination.CYBER_SALE)
            else -> {
                if (url.startsWith("/collection/")) {
                    AppLogger.debug("Navigate to collection: $url")
                } else {
                    AppLogger.debug("Unhandled navigation: $url")
                }
            }
        }
    }

    private fun parseCategoryURL(url: String) {
        val withoutScheme = url.replace("category://", "")
        val components = withoutScheme.split("?", limit = 2)
        val categoryName = URLDecoder.decode(components.firstOrNull() ?: "Category", StandardCharsets.UTF_8.name())

        var categoryId: String? = null
        var subCategoryId: String? = null
        var layoutType: String? = null
        val filters = mutableMapOf<String, String>()

        if (components.size > 1) {
            val queryString = components[1]
            val params = queryString.split("&")

            for (param in params) {
                val keyValue = param.split("=", limit = 2)
                if (keyValue.size != 2) continue

                val key = keyValue[0]
                val value = URLDecoder.decode(keyValue[1], StandardCharsets.UTF_8.name())

                when (key) {
                    "categoryId" -> categoryId = value
                    "subCategoryId" -> subCategoryId = value
                    "layout" -> layoutType = value
                    "filters" -> {
                        val filterPairs = value.split(",")
                        for (pair in filterPairs) {
                            val kv = pair.split(":", limit = 2)
                            if (kv.size == 2) {
                                filters[kv[0]] = kv[1]
                            }
                        }
                    }
                }
            }
        }

        _categoryNavigation.value = CategoryNavigationParams(
            categoryId = categoryId,
            categoryName = categoryName,
            subCategoryId = subCategoryId,
            filters = filters,
            layoutType = layoutType
        )
        setOverlay(OverlayDestination.CATEGORY_PAGE)
    }

    fun dismissCategoryPage() {
        if (_activeOverlay.value == OverlayDestination.CATEGORY_PAGE) {
            setOverlay(null)
        }
        _categoryNavigation.value = null
    }
}
